sherpa/utils/types.py

```diff
@@ -0,0 +1,55 @@
+#
+#  Copyright (C) 2024
+#  Smithsonian Astrophysical Observatory
+#
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License along
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+
+"""Useful types for Sherpa.
+
+This module should be considered an internal module as its contents is
+likely to change as types get added to Sherpa and the typing ecosystem
+in Python matures.
+
+"""
+
+from typing import Callable, Sequence, Union
+
+import numpy as np
+
+
+# Some of these may change to TypeAlias (once Python 3.10 is the
+# minimum) and then type/TypeAliasType once Python 3.12 is the
+# minimum.
+#
+
+# Try to be generic when using arrays as input or output. There is no
+# attempt to encode the data type or shape for ndarrays at this time.
+#
+ArrayType = Union[Sequence[float], np.ndarray]
+
+# Represent statistic evaluation, per-bin.
+#
+StatErrFunc = Callable[..., ArrayType]
+
+# Represent model evaluation. Using a Protocol may be better, but
+# for now keep with a Callable. Ideally the model would just
+# return an ndarray but
+#
+# - an ArithmeticConstantModel can return a scalar
+# - models could return a sequence rather than a ndarray
+#
+ModelFunc = Callable[..., ArrayType]
```

Comment:
Untortunately this should also allow returning a scalar value - *or* we change `ArithmeticConstantModel` to return an ndarray (we know the size to return). I shuld have a version where the return value includes a `float`, so need to check on why that isn't here.